### PR TITLE
Double-Decker Donut Disaster (Multi-Z Donutstation)

### LINKED ISSUE #49493: Donutstation's public mining shuttle cannot return to the station

### PR DESCRIPTION
**Pull request by MMMiracles**
## About The Pull Request
Adds a second floor to the Donutstation map.

## Why It's Good For The Game
The first ever multi-z station on /tg/station that also gives people reason to keep maintaining multi-z in the background past yearly one-off events.

Fixes #49493

## Changelog
:cl: MMMiracles (Donutstation)
add: Adds an entire second floor to Donutstation.
add: Multiple areas now have a second floor section, ranging from maintenance to various departments.
/:cl:

This things a fuckin' doozy with work put in on-off since October. I still have some work to do on maintenance and the Security wing, but it's far along enough for some criticism and nobody apparently looks at the spriting/mapping subforum anymore. It should be noted that the main goal of this is to set in stone a foundation for future department content updates an area that can freely developed on without worry of things like space restraint being a huge issue. A lot of this is built in mind with people accessing the 2nd floor through the departments themselves and maybe using the upper hall as a way to go around these rooms. I don't entirely want the 2nd floor to be as packed as the 1st because I don't want it taking away from the main section too much.

**Things so far**
- Basic mapping of the entire station's upper floor so it has a ceiling.
- The upper exterior is covered in various EVA catwalks and other miscellaneous bits to help fill in the space and keep it as visually interesting as a giant floor can be. 
- A secondary, 2-wide hallway that has a balcony overlooking the entire bottom hallway. This is mainly for walking access between 2nd story rooms. You can access this through a couple set of main hallway stairs located around the lower primary hallway.
- The gateway was moved up.
- The gravity generator was placed on a separate room on top of the engine room, with 3 various ladder access ways for engineers to enter.
- Vault and AI Upload have been moved upstairs with a command-only staircase access outside the Bridge. The tool storage was shifted slightly up to compensate and a new back maintenance path was added to fill in the spare room.
- EVA has a small upper section to store some extra materials.
- Cloning turned into a general treatment bay, added a small temporary morgue to hold bodies in a more thematic-sensible room instead of always sending them to the morgue.
- The AI Satellite was completely overhauled into a beefy two-story fort, now with an external glass bridge surrounding the base with the upper floor being a suspended platform that houses the 3 main AI locations.
- Maintenance has 3 major power rooms that transfer power and atmos between the floors, locked behind engineering access airlocks. There are also 2 similarly locked rooms that let you gain direct access to the upper floor's external hull.
- The Armory has been given a 2nd floor where lethal firearms are kept safe, which meant there was more room on the lower floor for non-lethal equipment to disperse to the crew. 
- The perma wing has been completely revamped to a 2-story prison wing to accommodate the prisoner role more.
- Added an upper brig area that connects to a balcony view between permabrig and the brig hallway. Has door access to the upper section of the armory.
- Chemistry's plumbing has been moved to a dedicated room above the Pharmacy with direct ladder access. No more walking through maintenance or other weird shit.
- Hydroponics has been given a 2nd floor with a basic apiary setup and a balcony overlook to the lower floor. Access from the remodeled backroom stairs.
- Library has been given a 2nd floor with a private study, printer room, and balcony overlook. You can access this floor through either the Curator's backroom ladder or the staircase in the game room.
- Cargo has had the mail room moved upstairs, along with a QM-access area that gives him a small balcony to overlook the warehouse as well as easier access to the mail room. There is a small request line outside this section for crew to request crates. The space where the mail room used to be downstairs was turned into a security checkpoint. The warehouse has a ladder leading to a 2nd floor with extra warehouse space for various warehouse activities.
- Atmospherics has a smaller upper section that is mostly for experimenting with pipes in a cleared out environment similar to plumbing. Comes with a mixing chamber, can be expanded upon with some effort and has some access to the lower atmos chambers if you put in some work.
- Dormitories has been modified greatly. The old, cramped dorm barracks was replaced with an arcade and the old shower/lock-room was replaced with a cubby with staircase access to the main segment of the living space. The upper floor dormitories has 2 large sleeping barracks and locker room/showers, gender-separated because Nanotrasen doesn't want no funny business on company time.

Mainly waiting for #49323 to be pushed, as the map changes are done and part from maybe some live tests is ready for review.

Feedback, feedback, FEEDBACK.

### CODE REVIEW

**Comment by actioninja** [comment]
oh great mmm took an upper decker on donut

**Comment by optimumtact** [comment]
gimme those fucking CANDY SHOTS

**Comment by carlarctg** [comment]
add a third zlevel, the frosting for the donut

**Comment by MMMiracles** [comment]
sHUT

**Comment by Qustinnus** [comment]
Give perma brig one of those cool catwalks that guards can use to look at prisoners and bully them

**Comment by Shadowflame909** [comment]
Can I fall through the floor

**Comment by MMMiracles** [comment]
yeah

**Comment by HomocidlePancakes** [comment]
Can I drop pianos on people to gib them?

**Comment by Shadowflame909** [comment]
Imagine falling through the floor to escape a murderboner.

Or falling through the floor into the armory.

This makes it harder for both antags and sec. I love it.

**Comment by Bawhoppen** [comment]
yes, but does it have original assets?
(I'm joking, good work)

**Comment by ArcaneMusic** [comment]
One thing that may need to get looked at is adding a second floor to telecoms, or a least just adding a relay to the second floor so that the radio works.

**Comment by Neo-0** [comment]
you can use it https://github.com/BeeStation/BeeStation-Hornet/pull/467

File: code/game/objects/structures/railing.dm
Comment (Qustinnus) [inline code/game/objects/structures/railing.dm]: if you're checking against dirs please use the flags because 9 doesn't mean anything to me. For example if this is north west do NORTH | WEST

Comment (spookydonut) [reply code/game/objects/structures/railing.dm]: NORTHWEST exists

**Comment by MCterra10** [comment]
oh god why must you do this

**Review by TheChosenEvilOne — COMMENTED** [review]
You don't need to use the baseturf helpers anywhere

File: _maps/donutstation.json
Comment (TheChosenEvilOne) [inline _maps/donutstation.json]:
```
"traits": [{"Up": 1}, {"Down": -1, "Baseturf": /turf/open/openspace}]
```
And remove all the baseturf helpers

Comment (MMMiracles) [reply _maps/donutstation.json]: This definitely doesn't work.

File: code/game/objects/structures/railing.dm
Comment (spookydonut) [inline code/game/objects/structures/railing.dm]:
```suggestion
	var/static/list/dir_NORTHWEST = list(NORTH, WEST) //ugly ass checks specific to the corner rails with intercardinal directions
```

File: code/game/objects/structures/railing.dm
Comment (spookydonut) [inline code/game/objects/structures/railing.dm]:
```suggestion
	if(dir == NORTHWEST && (get_dir(loc, target) in dir_NORTHWEST))
```

Comment (spookydonut) [reply code/game/objects/structures/railing.dm]: but honestly just do the proper bitwise operations since NORTHWEST == NORTH|WEST

Comment (spookydonut) [inline code/game/objects/structures/railing.dm]:
```
if(get_dir(loc, target) & dir)
	return !density
```

File: code/game/objects/structures/railing.dm
Comment (spookydonut) [reply code/game/objects/structures/railing.dm]: still must call parent

File: code/game/objects/structures/railing.dm
Comment (spookydonut) [inline code/game/objects/structures/railing.dm]:
```suggestion
/obj/structure/railing/corner/CanPass()
	SHOULD_CALL_PARENT(FALSE)
```

Comment (spookydonut) [reply code/game/objects/structures/railing.dm]: wait no, dont do this, just call ..() then return TRUE

File: _maps/donutstation.json
Comment (Jalleo) [inline _maps/donutstation.json]: https://github.com/tgstation/tgstation/issues/49009

**Comment by 81Denton** [comment]
I don't have time for a proper review, but this looks really, really good. I hope that you won't get discouraged by multi Z issues because having a working multi Z station map would be a huge step forward.

File: _maps/donutstation.json
Comment (Jalleo) [reply _maps/donutstation.json]: This issue is still existing

**Comment by spookydonut** [comment]
@MMMiracles merge conflict

**Comment by MMMiracles** [comment]
I know, I'm still working on stuff locally, plus a few other things I'm still waiting on before I can even think about getting it merged.

**Comment by PKPenguin321** [comment]
finna take an upper decker on this station

**Comment by MMMiracles** [comment]
rude ngl

**Comment by PKPenguin321** [comment]
Fuck, actioninja already made that joke nevermind

File: code/game/objects/structures/railing.dm
Comment (Fikou) [reply code/game/objects/structures/railing.dm]: does it still have to call parent?

Comment (Dennok) [reply code/game/objects/structures/railing.dm]: As i see now it dont call parrent.

**Comment by SpaceManiac** [comment]
Closing as WIP/stale. Contact a maintainer when work has continued.

**Comment by MMMiracles** [comment]
cobby is cute and likes to open closed prs

**Comment by Tharcoonvagh** [comment]
Are stairs constructible or destructible?
What about blob placement? Can you use the "go up/down" verbs to move as Bob overmind? 
And revs (or heads in a rev round) ? Do they lose upon changing zlevels on-station?
And teslas/singulos? Can they change levels?
Do explosions affect the upper level?
And the atmosphere?
Can you draw cult runes upstairs?
That's all atm

**Comment by 81Denton** [comment]
@Tharcoonvagh Make a forum thread for that please, that way the answers won't get lost in the comments of a map PR:
https://tgstation13.org/phpBB/viewforum.php?f=5

**Comment by Ty-the-Smonk** [comment]
Are we finally going to get a finished multi z map, and is the first one gonna be donut... there is some irony here.

File: code/game/atoms.dm
Comment (spookydonut) [inline code/game/atoms.dm]:
```suggestion
			return FALSE
```

File: code/game/atoms.dm
Comment (spookydonut) [inline code/game/atoms.dm]: `if(istype(get_area(T), /area/space))`

File: code/game/objects/structures/railings.dm
Comment (spookydonut) [inline code/game/objects/structures/railings.dm]: `return !(get_dir(loc, target) & dir)`

**Review by spookydonut — CHANGES_REQUESTED** [review]

**Comment by MMMiracles** [comment]
im WAITING ON ONE PR THATS BEEN SITTING FOR 17 DAYS 😡

**Comment by PKPenguin321** [comment]
make a 1 line comment only commit to tell the robot to fr*ck off

**Review by spookydonut — DISMISSED** [review]

**Comment by spookydonut** [comment]
@MMMiracles merge konflikt

**Comment by Dennok** [comment]
Perfectly. #49323 merged.

**Comment by Neo-0** [comment]
next is boxstation

**Comment by wesoda25** [comment]
awesome

**Pull request merged**
